requester/report.go

```diff
@@ -27,43 +31,52 @@ func (rep Report) String() string {
 // they are available and consumes them to build the report.
 // Returns the report when all the records have been collected.
 // Requester.collect will blocks until Requester.Records is empty.
-func (r *Requester) collect() Report {
+func (r *Requester) collect() (Report, error) {
+	recordsCap := defaultRecordsCap
+	if r.config.RunnerOptions.Requests > 0 {
+		recordsCap = r.config.RunnerOptions.Requests
+	}
+
 	rep := Report{
 		Config:  r.config,
-		Records: make([]Record, 0, r.config.RunnerOptions.Requests), // Provide capacity if known.
+		Records: make([]Record, 0, recordsCap), // Provide capacity if known.
 	}
 
-	for rec := range r.records {
+	for rec := range r.recordC {
+		select {
+		case err := <-r.errC:
+			return Report{}, err
```

Comment:
Why not call `cancel` here ? We stopped the process by the context with duration countdown is still going, and `errC` is not closed either.

Maybe we should not handle this here, but I think we should handle it anyway.
So either pass `cancel` to `collect` or read error inside `Run` and cancel before returning.

Reply:
`cancel` is already called in `Run`. It has nothing to do with error handling: we use `cancel` to close it once we're done we it to avoid leaks (similarly to `io.ReadCloser`s `Close`), so the right place to do it is just after calling `Dispatcher.Do(ctx, ...)`, regardless of how we handle the error it returns.